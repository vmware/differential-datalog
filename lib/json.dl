/* JSON parsing/serialization library.
 *
 * Functions in this library are bindings for functions in
 * the Rust `serde_json` crate.
 */

/* Deserialize an instance of type 'T from a string of JSON text.
 *
 * This conversion can fail if the structure of the input does not match the
 * structure expected by 'T, for example if 'T is a struct type but the input
 * contains something other than a JSON map.
 */
extern function from_json_string(json: string): Result<'T, string>

/* Serialize the given data structure as a String of JSON.
 *
 * Serialization can fail if 'T contains a map with non-string keys.
 */
extern function to_json_string(x: 'T): Result<string, string>

/* Represents any valid JSON value.
 */
#[rust="serde(from = \"__json::ValueWrapper\", into = \"__json::ValueWrapper\")"]
typedef JsonValue = // Represents a JSON null value.
                    JsonNull
                  | // Represents a JSON boolean.
                    JsonBool{b: bool}
                  | // Represents a JSON number, whether integer or floating point.
                    JsonNumber{n: JsonNum}
                  | // Represents a JSON string.
                    JsonString{s: istring}
                  | // Represents a JSON array.
                    JsonArray{a: Vec<JsonValue>}
                  | // Represents a JSON object.
                    JsonObject{o: Map<istring, JsonValue>}

/* Get JsonValue value as `bool`.  Returns `None` if value is not `JsonBool`.
 */
function jval_as_bool(v: JsonValue): Option<bool> =
{
    match (v) {
        JsonBool{b} -> Some{b},
        _ -> None
    }
}

/* Get JsonValue value as `bool`.  Returns `def` if value is not `JsonBool`.
 */
function jval_as_bool_or(v: JsonValue, def: bool): bool =
{
    match (v) {
        JsonBool{b} -> b,
        _ -> def
    }
}

/* Get JsonValue value as `JsonNum`.  Returns `None` if value is not `JsonNumber`.
 */
function jval_as_number(v: JsonValue): Option<JsonNum> =
{
    match (v) {
        JsonNumber{n} -> Some{n},
        _ -> None
    }
}

/* Get JsonValue value as `JsonNum`.  Returns `def` if value is not `JsonNumber`.
 */
function jval_as_number_or(v: JsonValue, def: JsonNum): JsonNum =
{
    match (v) {
        JsonNumber{n} -> n,
        _ -> def
    }
}

/* Get JsonValue value as string.  Returns `None` if value is not `JsonString`.
 */
function jval_as_string(v: JsonValue): Option<istring> =
{
    match (v) {
        JsonString{s} -> Some{s},
        _ -> None
    }
}

/* Get JsonValue value as string.  Returns `def` if value is not `JsonString`.
 */
function jval_as_string_or(v: JsonValue, def: istring): istring =
{
    match (v) {
        JsonString{s} -> s,
        _ -> def
    }
}

/* Get JsonValue value as array.  Returns `None` if value is not `JsonArray`.
 */
function jval_as_array(v: JsonValue): Option<Vec<JsonValue>> =
{
    match (v) {
        JsonArray{a} -> Some{a},
        _ -> None
    }
}

/* Get JsonValue value as array.  Returns `def` if value is not `JsonArray`.
 */
function jval_as_array_or(v: JsonValue, def: Vec<JsonValue>): Vec<JsonValue> =
{
    match (v) {
        JsonArray{a} -> a,
        _ -> def
    }
}

/* Get JsonValue value as object.  Returns `None` if value is not `JsonObject`.
 */
function jval_as_object(v: JsonValue): Option<Map<istring, JsonValue>> =
{
    match (v) {
        JsonObject{o} -> Some{o},
        _ -> None
    }
}

/* Get JsonValue value as object.  Returns `def` if value is not `JsonObject`.
 */
function jval_as_object_or(v: JsonValue, def: Map<istring, JsonValue>): Map<istring, JsonValue> =
{
    match (v) {
        JsonObject{o} -> o,
        _ -> def
    }
}

/* Get attribute by name.  Returns `None` if value is not `JsonObject` or
 * does not have the specified attribute.
 */
function jval_get(v: JsonValue, attr: istring): Option<JsonValue> =
{
    match (v) {
        JsonObject{o} -> map_get(o, attr),
        _ -> None
    }
}

/* Get attribute by name.  Returns `def` if value is not `JsonObject` or
 * does not have the specified attribute.
 */
function jval_get_or(v: JsonValue, attr: istring, def: JsonValue): JsonValue =
{
    match (v) {
        JsonObject{o} -> option_unwrap_or(map_get(o, attr), def),
        _ -> def
    }
}

typedef JsonNum = // Integer number.  The value must be in the range between the
                  // smallest `s64` number and the largest `u64` number,
                  // inclusive.  Values outside of this range will be truncated
                  // during serialization.
                  JsonInt{i: s128}
                | // NaN and infinity are not valid JSON values.
                  // When serializing to JSON, such values will be replaced with 0.
                  JsonFloat{d: double}

/* Wrapper type that serializes its inner object into binary formats via JSON string.
 *
 * This is a workaround for a `serde` limitation that prevents certain data types from
 * being serialized or deserialized from non-self-describing data formats like bincode.
 * In particular, this is the case for data types annotated with JSON-related serde
 * attributes like `tag` and `flatten`.  In practice this means that types annotated for
 * serialization to JSON cannot be serialized to other formats.
 *
 * The workaround is to use different serialization logic when serializing to binary and
 * non-binary formats.  In the binary case, we want to serialize such types into a string
 * and then serialize the string as a field in binary data.  In the non-binary case, we
 * stick to normal serde behavior.  `JsonWrapper` implements this choice by checking the
 * `is_human_readable` property of the serializer.
 *
 * The current implementation has an important flaw: serde currently does not provide a
 * reliable way to determine if a serializer supports self-describing data.  We use
 * `is_human_readable` attribute as an approximation, which is not always correct: binary
 * formats can be self-describing (BSON, Pickle), conversely not all human-readable formats
 * are self-describing.  Nevertheless, the current design enables useful scenarios, like
 * transcoding JSON to bincode.
 */
#[custom_serde]
typedef JsonWrapper<'T> = JsonWrapper{x: 'T}
