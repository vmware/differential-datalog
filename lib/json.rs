pub fn json_from_json_string<'de, T: serde::Deserialize<'de>>(
    json: &'de String,
) -> std_Result<T, String> {
    res2std(serde_json::from_str::<'de>(&*json))
}

pub fn json_to_json_string<T: serde::Serialize>(x: &T) -> std_Result<String, String> {
    res2std(serde_json::to_string(x))
}

pub struct ValueWrapper(serde_json::value::Value);

impl serde::Serialize for ValueWrapper {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        if serializer.is_human_readable() {
            self.0.serialize(serializer)
        } else {
            serde_json::to_string(&self.0)
                .map_err(|e| serde::ser::Error::custom(e))?
                .serialize(serializer)
        }
    }
}

impl<'de> Deserialize<'de> for ValueWrapper {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        if deserializer.is_human_readable() {
            Ok(ValueWrapper(serde_json::Value::deserialize(deserializer)?))
        } else {
            Ok(ValueWrapper(
                serde_json::from_str::<serde_json::Value>(
                    String::deserialize(deserializer)?.as_ref(),
                )
                .map_err(|e| serde::de::Error::custom(e))?,
            ))
        }
    }
}

impl From<ValueWrapper> for json_JsonValue {
    fn from(x: ValueWrapper) -> Self {
        json_JsonValue::from(x.0)
    }
}

impl From<json_JsonValue> for ValueWrapper {
    fn from(x: json_JsonValue) -> Self {
        ValueWrapper(serde_json::Value::from(x))
    }
}

impl From<serde_json::value::Value> for json_JsonValue {
    fn from(x: serde_json::value::Value) -> Self {
        match x {
            serde_json::value::Value::Null => json_JsonValue::json_JsonNull,
            serde_json::value::Value::Bool(b) => json_JsonValue::json_JsonBool { b },
            serde_json::value::Value::Number(n) => json_JsonValue::json_JsonNumber {
                n: json_JsonNum::from(n),
            },
            serde_json::value::Value::String(s) => json_JsonValue::json_JsonString {
                s: internment_intern(&s),
            },
            serde_json::value::Value::Array(a) => {
                let v: Vec<json_JsonValue> =
                    a.into_iter().map(|v| json_JsonValue::from(v)).collect();
                json_JsonValue::json_JsonArray {
                    a: std_Vec::from(v),
                }
            }
            serde_json::value::Value::Object(o) => json_JsonValue::json_JsonObject {
                o: o.into_iter()
                    .map(|(k, v)| (internment_intern(&k), json_JsonValue::from(v)))
                    .collect(),
            },
        }
    }
}

impl From<json_JsonValue> for serde_json::value::Value {
    fn from(x: json_JsonValue) -> Self {
        match x {
            json_JsonValue::json_JsonNull => serde_json::value::Value::Null,
            json_JsonValue::json_JsonBool { b } => serde_json::value::Value::Bool(b),
            json_JsonValue::json_JsonNumber { n } => json_val_from_num(n),
            json_JsonValue::json_JsonString { s } => {
                serde_json::value::Value::String(internment_ival(&s).clone())
            }
            json_JsonValue::json_JsonArray { a } => serde_json::value::Value::Array(
                a.into_iter()
                    .map(|v| serde_json::value::Value::from(v))
                    .collect(),
            ),
            json_JsonValue::json_JsonObject { o } => serde_json::value::Value::Object(
                o.into_iter()
                    .map(|(k, v)| {
                        (
                            internment_ival(&k).clone(),
                            serde_json::value::Value::from(v),
                        )
                    })
                    .collect(),
            ),
        }
    }
}

impl From<serde_json::Number> for json_JsonNum {
    fn from(n: serde_json::Number) -> Self {
        if n.is_u64() {
            json_JsonNum::json_JsonInt {
                i: n.as_u64().unwrap() as i128,
            }
        } else if n.is_i64() {
            json_JsonNum::json_JsonInt {
                i: n.as_i64().unwrap() as i128,
            }
        } else if n.is_f64() {
            json_JsonNum::json_JsonFloat {
                d: OrderedFloat::from(n.as_f64().unwrap()),
            }
        } else {
            panic!("JsonNum::from::<Number>(): unknown number format: '{}'", n)
        }
    }
}

fn json_val_from_num(n: json_JsonNum) -> serde_json::value::Value {
    match n {
        json_JsonNum::json_JsonInt { i } => {
            if i >= 0 {
                serde_json::value::Value::Number(serde_json::Number::from(i as u64))
            } else {
                serde_json::value::Value::Number(serde_json::Number::from(i as i64))
            }
        }
        json_JsonNum::json_JsonFloat { d } => serde_json::Number::from_f64(*d)
            .map(serde_json::value::Value::Number)
            .unwrap_or_else(|| serde_json::value::Value::Null),
    }
}

/* Some JSON schemas represent integers as strings.  This module serializes/deserializes
 * integers or any other types that implement `FromStr` and `ToString` traits to/from strings.
 * To use this module, add the following annotation to the DDlog field that needs to be
 * deserialized from a string:
 *
 * ```
 * typedef U64FromString = U64FromString {
 *   x: string,
 *   #[rust="serde(with=\"serde_string\")"]
 *   y: u64
 * }
 * ```
 */
pub mod serde_string {
    use serde::de;
    use serde::Deserialize;
    use serde::Deserializer;
    use serde::Serialize;
    use serde::Serializer;

    pub fn serialize<S, T>(x: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
        T: std::string::ToString,
    {
        x.to_string().serialize(serializer)
    }

    pub fn deserialize<'de, D, T>(deserializer: D) -> Result<T, D::Error>
    where
        D: Deserializer<'de>,
        T: std::str::FromStr,
        <T as std::str::FromStr>::Err: std::fmt::Display,
    {
        T::from_str(&String::deserialize(deserializer)?).map_err(|e| de::Error::custom(e))
    }
}

impl<T: Serialize> Serialize for json_JsonWrapper<T> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        if serializer.is_human_readable() {
            self.x.serialize(serializer)
        } else {
            serde_json::to_string(&self.x)
                .map_err(|e| serde::ser::Error::custom(e))?
                .serialize(serializer)
        }
    }
}

impl<'de, T: DeserializeOwned> Deserialize<'de> for json_JsonWrapper<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        if deserializer.is_human_readable() {
            Ok(json_JsonWrapper {
                x: T::deserialize(deserializer)?,
            })
        } else {
            Ok(json_JsonWrapper {
                x: serde_json::from_str::<T>(String::deserialize(deserializer)?.as_ref())
                    .map_err(|e| serde::de::Error::custom(e))?,
            })
        }
    }
}
