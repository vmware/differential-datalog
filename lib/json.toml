[dependencies.serde_json]
version = "1.0"
