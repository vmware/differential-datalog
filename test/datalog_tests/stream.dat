echo Initialize ZipCodes.;

start;
insert ZipCodes(94022, "Los Altos"),
insert ZipCodes(94035, "Moffett Field"),
insert ZipCodes(94039, "Mountain View"),
insert ZipCodes(94085, "Sunnyvale"),
insert ZipCodes(94088, "Onizuka Afb"),
insert ZipCodes(94301, "Palo Alto"),
insert ZipCodes(94303, "East Palo Alto"),
commit dump_changes;

echo Execute a single query.;
start;
insert ZipQueries(94301),
commit dump_changes;

echo Multiple queries in a single transaction.;

start;
insert ZipQueries(94022),
insert ZipQueries(94303),
insert ZipQueries2(94022, 5),
insert ZipQueries2(94303, 6),

commit dump_changes;

echo Modify zip code table.;
start;

delete ZipCodes(94022, "Los Altos"),
insert ZipCodes(95115, "San Jose"),
insert ZipCodes(95128, "San Jose"),

commit;

echo Query the modified table.;

start;
insert ZipQueries(94022),
insert ZipQueries(94303),
insert ZipQueries(95115),
insert ZipQueries2(94022, 5),
insert ZipQueries2(94303, 6),
insert ZipQueries2(95115, 4),

commit dump_changes;


echo Modify and query the table in the same transaction.;
start;

delete ZipCodes(95128, "San Jose"),
insert ZipCodes(95140, "Mount Hamilton"),
insert ZipCodes(95037, "Morgan Hill"),

insert ZipQueries(94022),
insert ZipQueries(95128),
insert ZipQueries(95037),
insert ZipQueries(95140),
insert ZipQueries2(94022, 5),
insert ZipQueries2(95128, 4),
insert ZipQueries2(95037, 12),
insert ZipQueries2(95140, 10),

commit dump_changes;

start;

insert KVStore(1, "1"),
insert KVStore(1, "1.1"),
insert KVStore(2, "2"),
insert KVStore(2, "2.1"),
insert KVStore(3, "3"),

commit dump_changes;

echo Test streaming queries;

start;

insert KVStreamQuery(1),
insert KVStreamQuery(2),
insert KVStreamQuery(3),
insert KVStreamQuery(4),

commit dump_changes;

echo Test relational queries;

start;

insert KVRelQuery(1),
insert KVRelQuery(2),
insert KVRelQuery(3),
insert KVRelQuery(4),

commit dump_changes;

start;

delete KVRelQuery(1),
delete KVRelQuery(2),
delete KVRelQuery(3),
delete KVRelQuery(4),

commit dump_changes;
