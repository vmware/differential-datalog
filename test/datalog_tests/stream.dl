/* Test "stream" relations.  They should take no memory.  */

import json

typedef Object = Object {
    field: u128
}

/* Raw JSON chunks. */
input stream Chunk(json: string)

/* Output an error message for each chunk that failed to parse. */
output stream ChunkParseError(err: string)
ChunkParseError(err) :- ParsedChunk(Err{err}).

/* Contains the result of parsing a chunk. */
stream ParsedChunk(data: Result<Object, string>)

ParsedChunk(from_json_string(json)) :- Chunk(json).

/* Successfully parsed chunks are stored in the Objects
 * relation. */
stream Objects(chunk: Object)
Objects(objs) :- ParsedChunk(Ok{objs}).

/* Convert chunks back to JSON. */
output stream CompressedChunk(json: string)
CompressedChunk(json) :-
    Objects(objs),
    Ok{var json} = to_json_string(objs).
