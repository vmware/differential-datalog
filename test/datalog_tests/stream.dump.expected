Initialize ZipCodes.
Execute a single query.
ZipResultsRS:
ZipResultsRS{.zip = 94301, .city = "Palo Alto"}: +1
ZipResultsSR:
ZipResultsSR{.zip = 94301, .city = "Palo Alto"}: +1
Multiple queries in a single transaction.
ZipResults2RS:
ZipResults2RS{.zip = 94022, .city = "Los Altos", .distance = 5}: +1
ZipResults2RS{.zip = 94303, .city = "East Palo Alto", .distance = 6}: +1
ZipResults2RSFiltered:
ZipResults2RSFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResults2RSRFiltered:
ZipResults2RSRFiltered{.zip = 94022, .city = "Los Altos", .distance = 5, .descr = "94022 is in Los Altos, 5 miles away"}: +1
ZipResults2RSRFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResults2SR:
ZipResults2SR{.zip = 94022, .city = "Los Altos", .distance = 5}: +1
ZipResults2SR{.zip = 94303, .city = "East Palo Alto", .distance = 6}: +1
ZipResults2SRFiltered:
ZipResults2SRFiltered{.zip = 94022, .city = "Los Altos", .distance = 5, .descr = "94022 is in Los Altos, 5 miles away"}: +1
ZipResults2SRFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResultsRS:
ZipResultsRS{.zip = 94022, .city = "Los Altos"}: +1
ZipResultsRS{.zip = 94303, .city = "East Palo Alto"}: +1
ZipResultsSR:
ZipResultsSR{.zip = 94022, .city = "Los Altos"}: +1
ZipResultsSR{.zip = 94303, .city = "East Palo Alto"}: +1
Modify zip code table.
Query the modified table.
ZipResults2RS:
ZipResults2RS{.zip = 94303, .city = "East Palo Alto", .distance = 6}: +1
ZipResults2RS{.zip = 95115, .city = "San Jose", .distance = 4}: +1
ZipResults2RSFiltered:
ZipResults2RSFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResults2RSFiltered{.zip = 95115, .city = "San Jose", .distance = 4, .descr = "95115 is in San Jose, 4 miles away"}: +1
ZipResults2RSRFiltered:
ZipResults2RSRFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResults2SR:
ZipResults2SR{.zip = 94303, .city = "East Palo Alto", .distance = 6}: +1
ZipResults2SR{.zip = 95115, .city = "San Jose", .distance = 4}: +1
ZipResults2SRFiltered:
ZipResults2SRFiltered{.zip = 94303, .city = "East Palo Alto", .distance = 6, .descr = "94303 is in East Palo Alto, 6 miles away"}: +1
ZipResultsRS:
ZipResultsRS{.zip = 94303, .city = "East Palo Alto"}: +1
ZipResultsRS{.zip = 95115, .city = "San Jose"}: +1
ZipResultsSR:
ZipResultsSR{.zip = 94303, .city = "East Palo Alto"}: +1
ZipResultsSR{.zip = 95115, .city = "San Jose"}: +1
Modify and query the table in the same transaction.
ZipResults2RS:
ZipResults2RS{.zip = 95037, .city = "Morgan Hill", .distance = 12}: +1
ZipResults2RS{.zip = 95140, .city = "Mount Hamilton", .distance = 10}: +1
ZipResults2RSFiltered:
ZipResults2RSFiltered{.zip = 95037, .city = "Morgan Hill", .distance = 12, .descr = "95037 is in Morgan Hill, 12 miles away"}: +1
ZipResults2RSFiltered{.zip = 95140, .city = "Mount Hamilton", .distance = 10, .descr = "95140 is in Mount Hamilton, 10 miles away"}: +1
ZipResults2RSRFiltered:
ZipResults2RSRFiltered{.zip = 95037, .city = "Morgan Hill", .distance = 12, .descr = "95037 is in Morgan Hill, 12 miles away"}: +1
ZipResults2RSRFiltered{.zip = 95140, .city = "Mount Hamilton", .distance = 10, .descr = "95140 is in Mount Hamilton, 10 miles away"}: +1
ZipResults2SR:
ZipResults2SR{.zip = 95037, .city = "Morgan Hill", .distance = 12}: +1
ZipResults2SR{.zip = 95140, .city = "Mount Hamilton", .distance = 10}: +1
ZipResults2SRFiltered:
ZipResults2SRFiltered{.zip = 95037, .city = "Morgan Hill", .distance = 12, .descr = "95037 is in Morgan Hill, 12 miles away"}: +1
ZipResults2SRFiltered{.zip = 95140, .city = "Mount Hamilton", .distance = 10, .descr = "95140 is in Mount Hamilton, 10 miles away"}: +1
ZipResultsRS:
ZipResultsRS{.zip = 95037, .city = "Morgan Hill"}: +1
ZipResultsRS{.zip = 95140, .city = "Mount Hamilton"}: +1
ZipResultsSR:
ZipResultsSR{.zip = 95037, .city = "Morgan Hill"}: +1
ZipResultsSR{.zip = 95140, .city = "Mount Hamilton"}: +1
Test streaming queries
KVStreamResponse:
KVStreamResponse{.k = 1, .v = "1"}: +1
KVStreamResponse{.k = 1, .v = "1.1"}: +1
KVStreamResponse{.k = 2, .v = "2"}: +1
KVStreamResponse{.k = 2, .v = "2.1"}: +1
KVStreamResponse{.k = 3, .v = "3"}: +1
Test relational queries
KVRelResponse:
KVRelResponse{.k = 1, .v = "1"}: +1
KVRelResponse{.k = 1, .v = "1.1"}: +1
KVRelResponse{.k = 2, .v = "2"}: +1
KVRelResponse{.k = 2, .v = "2.1"}: +1
KVRelResponse{.k = 3, .v = "3"}: +1
KVRelResponse:
KVRelResponse{.k = 1, .v = "1"}: -1
KVRelResponse{.k = 1, .v = "1.1"}: -1
KVRelResponse{.k = 2, .v = "2"}: -1
KVRelResponse{.k = 2, .v = "2.1"}: -1
KVRelResponse{.k = 3, .v = "3"}: -1
ControlledStreamOfSums:
ControlledStreamOfSums{.k = "a", .s = 3}: +1
ControlledStreamOfSums{.k = "b", .s = 12}: +1
StreamGroups:
StreamGroups{.k = "a", .vs = [1, 2]}: +1
StreamGroups{.k = "b", .vs = [3, 4, 5]}: +1
StreamGroups{.k = "c", .vs = [6]}: +1
StreamOfSums:
StreamOfSums{.k = "a", .s = 3}: +1
StreamOfSums{.k = "b", .s = 12}: +1
StreamOfSums{.k = "c", .s = 6}: +1
ControlledStreamOfSums:
ControlledStreamOfSums{.k = "a", .s = 3}: -1
ControlledStreamOfSums{.k = "a", .s = 10}: +1
ControlledStreamOfSums{.k = "b", .s = 12}: -1
StreamGroups:
StreamGroups{.k = "a", .vs = [7]}: +1
StreamGroups{.k = "b", .vs = [8]}: +1
StreamGroups{.k = "c", .vs = [9]}: +1
StreamGroups{.k = "d", .vs = [10]}: +1
StreamOfSums:
StreamOfSums{.k = "a", .s = 3}: -1
StreamOfSums{.k = "a", .s = 10}: +1
StreamOfSums{.k = "b", .s = 12}: -1
StreamOfSums{.k = "b", .s = 20}: +1
StreamOfSums{.k = "c", .s = 6}: -1
StreamOfSums{.k = "c", .s = 15}: +1
StreamOfSums{.k = "d", .s = 10}: +1
ControlledStreamOfSums:
ControlledStreamOfSums{.k = "f", .s = 25}: +1
StreamGroups:
StreamGroups{.k = "e", .vs = [11]}: +1
StreamGroups{.k = "f", .vs = [12, 13]}: +1
StreamOfSums:
StreamOfSums{.k = "e", .s = 11}: +1
StreamOfSums{.k = "f", .s = 25}: +1
ControlledStreamOfSums:
ControlledStreamOfSums{.k = "a", .s = 10}: -1
ControlledStreamOfSums:
ControlledStreamOfSums{.k = "a", .s = 300}: +1
StreamGroups:
StreamGroups{.k = "a", .vs = [100, 200]}: +1
StreamOfSums:
StreamOfSums{.k = "a", .s = 10}: -1
StreamOfSums{.k = "a", .s = 310}: +1
Clock: 1000
TimeWindow:
TimeWindow{.start = 995, .end = 1000}: +1
TimeWindow{.start = 1000, .end = 1005}: +1
Data1
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 9}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 27}: +1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12]}: +1
Data2
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 9}: -1
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 56}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 27}: -1
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 337}: +1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12]}: -1
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 100, 200]}: +1
Clock: 1001
Data3
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 56}: -1
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 71}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 337}: -1
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 647}: +1
Clock: 1005
TimeWindow:
TimeWindow{.start = 995, .end = 1000}: -1
TimeWindow{.start = 1005, .end = 1010}: +1
Data4
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 69}: +1
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 71}: -1
DataAvg{.window = TimeWindow{.start = 1005, .end = 1010}, .avg = 38}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 647}: -1
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 762}: +1
DataSum{.window = TimeWindow{.start = 1005, .end = 1010}, .sum = 76}: +1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 15, 100, 200]}: +1
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 100, 200]}: -1
DataUnique{.window = TimeWindow{.start = 1005, .end = 1010}, .unique_vals = [16, 60]}: +1
Clock: 1007
Data5
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 69}: -1
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 96}: +1
DataAvg{.window = TimeWindow{.start = 1005, .end = 1010}, .avg = 38}: -1
DataAvg{.window = TimeWindow{.start = 1005, .end = 1010}, .avg = 475}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 762}: -1
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 1162}: +1
DataSum{.window = TimeWindow{.start = 1005, .end = 1010}, .sum = 76}: -1
DataSum{.window = TimeWindow{.start = 1005, .end = 1010}, .sum = 2376}: +1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 15, 100, 200]}: -1
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 15, 100, 200, 400]}: +1
DataUnique{.window = TimeWindow{.start = 1005, .end = 1010}, .unique_vals = [16, 60]}: -1
DataUnique{.window = TimeWindow{.start = 1005, .end = 1010}, .unique_vals = [16, 60, 600, 800, 900]}: +1
Clock: 1011
DataAvg:
DataAvg{.window = TimeWindow{.start = 1000, .end = 1005}, .avg = 96}: -1
DataSum:
DataSum{.window = TimeWindow{.start = 1000, .end = 1005}, .sum = 1162}: -1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1000, .end = 1005}, .unique_vals = [5, 10, 12, 15, 100, 200, 400]}: -1
TimeWindow:
TimeWindow{.start = 1000, .end = 1005}: -1
TimeWindow{.start = 1010, .end = 1015}: +1
Data5
DataAvg:
DataAvg{.window = TimeWindow{.start = 1005, .end = 1010}, .avg = 341}: +1
DataAvg{.window = TimeWindow{.start = 1005, .end = 1010}, .avg = 475}: -1
DataAvg{.window = TimeWindow{.start = 1010, .end = 1015}, .avg = 10}: +1
DataSum:
DataSum{.window = TimeWindow{.start = 1005, .end = 1010}, .sum = 2376}: -1
DataSum{.window = TimeWindow{.start = 1005, .end = 1010}, .sum = 2392}: +1
DataSum{.window = TimeWindow{.start = 1010, .end = 1015}, .sum = 10}: +1
DataUnique:
DataUnique{.window = TimeWindow{.start = 1005, .end = 1010}, .unique_vals = [8, 16, 60, 600, 800, 900]}: +1
DataUnique{.window = TimeWindow{.start = 1005, .end = 1010}, .unique_vals = [16, 60, 600, 800, 900]}: -1
DataUnique{.window = TimeWindow{.start = 1010, .end = 1015}, .unique_vals = [10]}: +1
